### PR TITLE
scipy style sparse ndarray constructor and misc sparse fixes

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
- fix wrong usage of rowsparsepull in sparse LR example
- update csr_matrix / row_sparse_array interface like scipy
- misc doc updates

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated.
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Intersting edge cases to note here

### CODE REVIEW

File: python/mxnet/ndarray/sparse.py [hunk 1]
```diff
@@ -711,47 +719,88 @@ def copyto(self, other):
...
+    """check s1 == s2 if both are not None"""
+    if s1 and s2 and s1 != s2:
+        raise ValueError("Shape mismatch detected. " + str(s1) + " v.s. " + str(s2))
+
+def csr_matrix(arg1, shape=None, ctx=None, dtype=None):
+    """Creates a `CSRNDArray`, an 2D array with compressed sparse row (CSR) format.
+
+    The CSRNDArray can be instantiated in several ways:
+
+    - csr_matrix(D):
```
Comment (piiswrong) [inline python/mxnet/ndarray/sparse.py]: Have you checked if this will render correctly?

File: python/mxnet/ndarray/sparse.py [hunk 2]
```diff
@@ -916,17 +1077,15 @@ def zeros(stype, shape, ctx=None, dtype=None, aux_types=None, **kwargs):
...
-    if aux_types is None:
-        if stype == 'row_sparse' or stype == 'csr':
-            aux_types = _STORAGE_AUX_TYPES[stype]
-        else:
-            raise Exception("unknown storage type")
-    assert(len(aux_types) == len(_STORAGE_AUX_TYPES[stype]))
+    if stype == 'row_sparse' or stype == 'csr':
+        aux_types = _STORAGE_AUX_TYPES[stype]
+    else:
+        raise Exception("unknown storage type")
```
Comment (piiswrong) [inline python/mxnet/ndarray/sparse.py]: include stype in the error

File: python/mxnet/ndarray/sparse.py (hunk 1 above)
Comment (eric-haibin-lin) [reply python/mxnet/ndarray/sparse.py]: Yes the rendering is correct. I'll post a preview link soon

**Comment by eric-haibin-lin** [comment]
Preview: http://34.211.112.96/api/python/ndarray/sparse.html#mxnet.ndarray.sparse.csr_matrix

File: python/mxnet/ndarray/sparse.py [hunk 3]
```diff
@@ -711,47 +719,88 @@ def copyto(self, other):
...
+            raise TypeError('values must be array like object')
+    return source_array
+
+def _check_shape(s1, s2):
+    """check s1 == s2 if both are not None"""
+    if s1 and s2 and s1 != s2:
+        raise ValueError("Shape mismatch detected. " + str(s1) + " v.s. " + str(s2))
+
+def csr_matrix(arg1, shape=None, ctx=None, dtype=None):
+    """Creates a `CSRNDArray`, an 2D array with compressed sparse row (CSR) format.
```
Comment (szha) [inline python/mxnet/ndarray/sparse.py]: 'an 2D array' -> 'a 2D-array'

File: python/mxnet/ndarray/sparse.py [hunk 4]
```diff
@@ -830,30 +957,65 @@ def row_sparse_array(data, indices, shape, ctx=None, dtype=None, indices_type=No
...
            [ 0.,  0.],
            [ 3.,  4.],
            [ 0.,  0.]], dtype=float32)
+
+    See Also
+    --------
+    RowSparseNDArray : MXNet NDArray in row sparse format.
     """
+    # construct a row sparse array from (D0, D1 ..) or (data, indices)
+    if isinstance(arg1, tuple):
```
Comment (szha) [inline python/mxnet/ndarray/sparse.py]: list and tuple?

Comment (eric-haibin-lin) [reply python/mxnet/ndarray/sparse.py]: this is intentional. otherwise there're too many cases since data and indices can be list, too

**Pull request merged**